Fixes to cell type tables

allyhawkins opened the pull request:
Closes #512 
Closes #527 

This PR makes the necessary adjustments to fix some of the formatting and display of the tables that are mentioned in #512. 

- In the `create_celltype_df` function, I added a check for submitter cell types and set NA to "Unclassified cell". 
- I also made sure to add in the re-leveling so that the table for submitter cell types is in order of percentage of cells with "Unclassified cell" last. 
- I also fixed a small typo that meant we weren't setting the cell ontology labels as the annotations, which is what we want to do. That should now fix the situation of seeing the cell ontology ids as the axis labels in the plots. 
- The last thing I did was just add a check that all submitter cell types are not NA. If that's the case, then `has_submitter` is set to `FALSE`. No more blank UMAPs or heatmaps. 

Note that because I'm running CellAssign right now I haven't been able to fully test this with Nextflow, but I did test this locally.

sjspielman commented:
> Note that because I'm running CellAssign right now I haven't been able to fully test this with Nextflow, but I did test this locally.

Do you have a locally-run report you can attach here?

> The last thing I did was just add a check that all submitter cell types are not NA. If that's the case

Just generally thinking about when/how this would happen in the first place. The circumstance I can think of is if we end up filtering out _all_ the cells that the submitter annotated, so only `NA` labels are left in the processed object. Are there other situations I'm not immediately seeing?

sjspielman commented on templates/qc_report/main_qc_report.rmd:
```diff
@@ -114,7 +114,8 @@ if (has_processed) {
   has_singler <- "singler" %in% metadata(processed_sce)$celltype_methods
   has_cellassign <- "cellassign" %in% metadata(processed_sce)$celltype_methods
-  has_submitter <- "submitter" %in% metadata(processed_sce)$celltype_methods
+  has_submitter <- "submitter" %in% metadata(processed_sce)$celltype_methods &
```
👍

sjspielman commented on templates/qc_report/utils/celltype_functions.rmd:
```diff
@@ -29,10 +29,22 @@ create_celltype_df <- function(processed_sce) {
     )
+  
+  if("submitter_celltype_annotation" %in% names(celltype_df)){
+    celltype_df <- celltype_df |> 
+      dplyr::mutate(submitter_celltype_annotation = dplyr::if_else(
+        is.na(submitter_celltype_annotation),
+        "Unclassified cell", 
+        submitter_celltype_annotation
+      ) |>
+        forcats::fct_infreq() |>
+        forcats::fct_relevel("Unclassified cell", after = Inf)
+      )
```
I wonder if we should modify and use the `prepare_annotation_values()` function. One way to do that is below (also I un-hardcoded some strings as part of it). 

What do you think? 

We could still take this conceptual approach without the specific way I have implemented it; `fct_relevel()` for example will work even if provided a string that isn't present but it will toss out a warning which will need to be suppressed. Using `fct_string` means no warning will be applied.

```
prepare_annotation_values <- function(df, annotation_column, is_submitter) {

  submitter_string <- "Unclassified cell" # use for unclassified submitter cells 
  method_string  <- "Unknown cell type" # use for unclassified singler and cellassign cells
  fct_string <- ifelse(is_submitter, submitter_string, method_string) # which string to consider when reordering the factor

  df |>
    dplyr::mutate(
      {{ annotation_column }} := dplyr::case_when(
        # submitter condition - HAS TO BE FIRST since these are also NA which is the singler check
        is_submitter  ~ submitter_string,
        # singler condition
        is.na({{ annotation_column }})~ method_string,
        # cellassign conditon
        {{ annotation_column }} == "other" ~ method_string,
        # otherwise, keep it
        .default = {{ annotation_column }}
      ) |>
        # order column by number of cells
        forcats::fct_infreq() |>
        # relevel so the applied unknown is at the end
        forcats::fct_relevel(fct_string, after = Inf)
       ## OR this will work but with a warning so have to suppress them from chunk
        forcats::fct_relevel(c(submitter_string, method_string), after = Inf)
    )
```

sjspielman replied:
Oh, and can default `is_submitter=FALSE` in the function, and set to true when calling submitters.

sjspielman commented on templates/qc_report/celltypes_qc.rmd:
```diff
@@ -132,7 +132,7 @@ glue::glue(
 glue::glue(
   "* Cell types were annotated using [`SingleR`](https://bioconductor.org/packages/release/bioc/html/SingleR.html), which uses a reference-based approach ([Looney _et al._ 2019](https://doi.org/10.1038/s41590-018-0276-y)).
-  The `{metadata(sce)$singler_reference}` reference dataset, obtained from the [`celldex` package](http://bioconductor.org/packages/release/data/experiment/html/celldex.html) package, was used for cell typing."
+  The `{metadata(processed_sce)$singler_reference}` reference dataset, obtained from the [`celldex` package](http://bioconductor.org/packages/release/data/experiment/html/celldex.html) package, was used for cell typing."
```
we [missed this](https://en.wikipedia.org/wiki/Major_Major_Major_Major) before 
```suggestion
  The `{metadata(processed_sce)$singler_reference}` reference dataset, obtained from the [`celldex` package](http://bioconductor.org/packages/release/data/experiment/html/celldex.html), was used for cell typing."
```

sjspielman reviewed (commented):
A couple small comments from looking at the report you posted in https://github.com/AlexsLemonade/scpca-nf/pull/529#issuecomment-1783234093 . One I left as a suggestion (too many "package"s), but the other I can't comment on directly. 

On L143, we have:
```
 Marker genes associated with `{metadata(processed_sce)$cellassign_reference}` tissue were obtained from [PanglaoDB](https://panglaodb.se/).
```

But `{metadata(processed_sce)$cellassign_reference}` still contains the `Panglao-` prefix. We need to remove that for pretty printing purposes, but I wonder where we should do that! Can either be in the report just for printing, or back in `add_celltypes_to_sce.R` to change how we save it in the final SCE.

allyhawkins commented:
> Just generally thinking about when/how this would happen in the first place. The circumstance I can think of is if we end up filtering out all the cells that the submitter annotated, so only NA labels are left in the processed object. Are there other situations I'm not immediately seeing?

Replying to this comment first, while I go through the other comments. 
I added a check for not all submitter cell types being NA, because some libraries in SCPCP000005 were never annotated in the first place. Because submitter cell types are added on a project level (we have one metadata file for cell types per project) all the cell types will be assigned as NA for a library that the submitter never annotated.

sjspielman commented:
> Because submitter cell types are added on a project level

Project level! Yes, thanks!

allyhawkins commented:
> But {metadata(processed_sce)$cellassign_reference} still contains the Panglao- prefix. We need to remove that for pretty printing purposes, but I wonder where we should do that! Can either be in the report just for printing, or back in add_celltypes_to_sce.R to change how we save it in the final SCE.

I updated this to extract the actual tissue type from the name, removing the PanglaoDB in https://github.com/AlexsLemonade/scpca-nf/pull/528/commits/c57e64c2a0c63e8b686fea0d155fd09cd0b77e89. 

> I wonder if we should modify and use the prepare_annotation_values() function. One way to do that is below (also I un-hardcoded some strings as part of it).

Thanks for the suggestion here! I updated the function to consider submitter annotations rather than directly modifying them in `create_celltype_df`. I did test to make sure that everything was named correctly after implementing those changes in https://github.com/AlexsLemonade/scpca-nf/pull/528/commits/b734f96f75da15b641055c1ac914f1e22b3ad5b6. 

This should be ready for another round of review @sjspielman.

sjspielman commented on templates/qc_report/celltypes_qc.rmd:
````diff
@@ -132,15 +132,21 @@ glue::glue(
 
 
 ```{r, eval = has_cellassign, results='asis'}
+tissue_type <- stringr::word(
+  metadata(processed_sce)$cellassign_reference,
+  -1,
+  sep = "-"
+)
+
````
I'm of two minds here... (although really maybe i am of one mind and playing devil's advocate with the other).

1. This is nicely general if we use other references besides PanglaoDB one day. We'll want to add some documentation that ensures future references are of the format `<databasename>-<tissuename>` moving forward, somewhere here. I envision a _super quick_ addition of "if/when we add future references use format `<databasename>-<tissuename>`".
https://github.com/AlexsLemonade/scpca-nf/blob/4c6d9a24469bcd7c069ecf02e74d4f314984a369/internal-instructions.md?plain=1#L167-L174

3. That said, _do we_ anticipate using something besides PanglaoDB? If we absolutely really really don't, then this could be simplified to a `str_remove()`. But, I am more in favor of keeping what you have here since I like to keep things flexible!

allyhawkins replied:
I am somewhat hesitant to add the documentation in at this point, since I envision changes in the cell type reference file naming as part of #470. So I think I'll just comment on that issue with a link to the line here that I altered. We may need to make further adjustments anyways.

sjspielman replied:
very reasonable!

sjspielman reviewed (approved):
LGTM!

I left one main comment about some docs we might want to add given some reasonable coding choices you made. If you disagree with my thoughts, let me know! The punchline of what I'm (probably) recommending means you'll want to add a little more docs in `internal-instructions.md`, but I don't need to necessarily see again unless you want to tag me back in!
